Running RAGAS behind a proxy

[x] I have checked the [documentation](https://docs.ragas.io/) and related resources and couldn't resolve my bug.

**Describe the bug**
I am using Ragas with in-house openAI API. When I am using it directly, everything is fine. However, when I am trying  to use it through our proxy, I encountered the infamous APIConnectionError. I have to specify that when I am using the API directly with the ChatOpenAI instance, everything is fine, with ou without the proxy (same thing for the proxy). So, the issue is really at the Ragas level

Ragas version: 0.3.8
Python version: 3.12

**Code to Reproduce**
```python
import httpx
import nest_asyncio
from langchain_openai import ChatOpenAI
from langchain_core.embeddings import Embeddings
from ragas import evaluate, EvaluationDataset
from ragas.metrics import answer_correctness

client = httpx.Client(proxy="http://user:passwd@proxy-server:8080")

agent = ChatOpenAI(
        base_url=config.agent.url,
        api_key=config.session.token,
        model=config.agent.model_name,
        http_client=client
)

# The agent run perfectly with the proxy
answer = agent.invoke("Answer 'pong' if you get a 'ping', answer 'crash' otherwise:\n'ping'")

# I cannot provide the exact embedder code, but it is not the culprit here
embedder = Embeddings()


nest_asyncio.apply()

evaluation = evaluate(
                dataset,
                metrics=[answer_correctness],
                llm=agent,
                embeddings=embedder,
                allow_nest_asyncio=True,
                raise_exceptions=True,
                show_progress=False,
)
```

**Error trace**
```
@contextlib.contextmanager
    def map_httpcore_exceptions() -> typing.Iterator[None]:
        global HTTPCORE_EXC_MAP
        if len(HTTPCORE_EXC_MAP) == 0:
            HTTPCORE_EXC_MAP = _load_httpcore_exceptions()
        try:
>           yield
...
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

map = {<class 'TimeoutError'>: <class 'httpcore.ConnectTimeout'>, <class 'OSError'>: <class 'httpcore.ConnectError'>, <class 'anyio.BrokenResourceError'>: <class 'httpcore.ConnectError'>}

    @contextlib.contextmanager
    def map_exceptions(map: ExceptionMapping) -> typing.Iterator[None]:
        try:
            yield
        except Exception as exc:  # noqa: PIE786
            for from_exc, to_exc in map.items():
                if isinstance(exc, from_exc):
>                   raise to_exc(exc) from exc
E                   httpcore.ConnectError: All connection attempts failed

..\.venv\Lib\site-packages\httpcore\_exceptions.py:14: ConnectError
The above exception was the direct cause of the following exception:

self = <openai.AsyncOpenAI object at 0x000001F79C995160>
cast_to = <class 'openai.types.chat.chat_completion.ChatCompletion'>
options = FinalRequestOptions(method='post', url='/chat/completions', params={}, headers={'X-Stainless-Raw-Response': 'true'}, m...s."\n}\nOutput: ', 'role': 'user'}], 'model': 'gemma3', 'n': 1, 'stream': False, 'temperature': 0.01}, extra_json=None)

    async def request(
        self,
        cast_to: Type[ResponseT],
        options: FinalRequestOptions,
        *,
        stream: bool = False,
        stream_cls: type[_AsyncStreamT] | None = None,
    ) -> ResponseT | _AsyncStreamT:
        if self._platform is None:
            # `get_platform` can make blocking IO calls so we
            # execute it earlier while we are in an async context
            self._platform = await asyncify(get_platform)()
    
        cast_to = self._maybe_override_cast_to(cast_to, options)
    
        # create a copy of the options we were given so that if the
        # options are mutated later & we then retry, the retries are
        # given the original options
        input_options = model_copy(options)
        if input_options.idempotency_key is None and input_options.method.lower() != "get":
            # ensure the idempotency key is reused between requests
            input_options.idempotency_key = self._idempotency_key()
    
        response: httpx.Response | None = None
        max_retries = input_options.get_max_retries(self.max_retries)
    
        retries_taken = 0
        for retries_taken in range(max_retries + 1):
            options = model_copy(input_options)
            options = await self._prepare_options(options)
    
            remaining_retries = max_retries - retries_taken
            request = self._build_request(options, retries_taken=retries_taken)
            await self._prepare_request(request)
    
            kwargs: HttpxSendArgs = {}
            if self.custom_auth is not None:
                kwargs["auth"] = self.custom_auth
    
            if options.follow_redirects is not None:
                kwargs["follow_redirects"] = options.follow_redirects
    
            log.debug("Sending HTTP Request: %s %s", request.method, request.url)
    
            response = None
            try:
>               response = await self._client.send(
                    request,
                    stream=stream or self._should_stream_response_body(request=request),
                    **kwargs,
                )

..\.venv\Lib\site-packages\openai\_base_client.py:1529: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
...
                log.debug("Raising connection error")
>               raise APIConnectionError(request=request) from err
E               openai.APIConnectionError: Connection error.

..\.venv\Lib\site-packages\openai\_base_client.py:1561: APIConnectionError
```

**Expected behavior**
The evaluate() function has to manage proxy connection. Would it change something if the proxy parameters are set through environment variables, such as:
```
os.environ["HTTP_PROXY"] = "http://proxy-server:8080"
os.environ["HTTPS_PROXY"] = "http://proxy-server:8080"
```

I have change the code to follow the [documentation](https://docs.ragas.io/en/stable/concepts/metrics/available_metrics/answer_correctness/?h=#answer-correctness):
```python
from httpx import AsyncClient
from ragas.metrics.collections import AnswerCorrectness
from ragas.llms import llm_factory
from ragas.embeddings.base import embedding_factory

from core.models.query import Query

class RagasAnswerCorrectnessScorer:
    """
     Management of the RAGAS answer correctness score calculation
    """

    def __init__(self, http_client=AsyncClient):
        self.name = 'answer_correctness'

        self.llm = llm_factory(
            model='gemma3',
            provider="litellm",
            base_url="https://api.lite.llm/v1/",
            client=http_client)

        self.embedder = embedding_factory(
            model='nomic-embed-text',
            provider="litellm",
            base_url="https://api.lite.llm/v1/embeddings",
            client=http_client)


        self.scorer = AnswerCorrectness(llm=self.llm, embeddings=self.embedder)

    async def get(self, queries: list[Query]) -> list[Query]:
        """
        Calculates the answer correctness score for the whole dataset.
        """
        for q in range(len(queries)):
            result = await self.scorer.ascore(
                user_input=queries[q].user_input,
                response=queries[q].response,
                reference=queries[q].reference,
            )

            print(f"score: {result}")
            queries[q].set_score(self.name, result.value)

        return queries
```
When I invoke the scoring function:
````python
scorer = RagasAnswerCorrectnessScorer(http_client=client_provider.get_async_client())
scored = asyncio.run(scorer.get(queries))
```

I am ending up with a TypeError:
```python
self = InstructorLLM(provider='litellm', model='gemma3', client=<Instructor:sync>, temperature=0.01, max_tokens=1024, top_p=0.1, +1 more)
prompt = 'Given a question and an answer, analyze the complexity of each sentence in the answer. Break down each sentence into ...'
response_model = <class 'ragas.metrics.collections._answer_correctness.StatementGeneratorOutput'>

    async def agenerate(
        self,
        prompt: str,
        response_model: t.Type[InstructorTypeVar],
    ) -> InstructorTypeVar:
        """Asynchronously generate a response using the configured LLM."""
        messages = [{"role": "user", "content": prompt}]
    
        # If client is not async, raise a helpful error
        if not self.is_async:
>           raise TypeError(
                "Cannot use agenerate() with a synchronous client. Use generate() instead."
            )
E           TypeError: Cannot use agenerate() with a synchronous client. Use generate() instead.

..\.venv\Lib\site-packages\ragas\llms\base.py:731: TypeError
```
I am not master of asynchronous requests... where the issue could come from?

--- a/src/ragas/llms/base.py
+++ b/src/ragas/llms/base.py
@@ -805,11 +805,66 @@ class InstructorLLM(InstructorBaseRagasLLM):
         return google_kwargs
 
     def _check_client_async(self) -> bool:
-        """Determine if the client is async-capable."""
+        """Determine if the client is async-capable.
+
+        Handles multiple cases:
+        1. Instructor-wrapped AsyncInstructor clients (OpenAI/Anthropic/etc)
+        2. Instructor-wrapped Instructor clients that wrap async underlying clients
+        3. Direct async clients with chat.completions.create
+        4. Instructor-wrapped clients where the underlying client is in a closure
+        """
         try:
+            # Check if this is an AsyncInstructor wrapper
+            if self.client.__class__.__name__ == "AsyncInstructor":
+                return True
+
+            # Check if this is a sync Instructor wrapper that wraps an async client
+            if hasattr(self.client, "client"):
+                underlying = self.client.client
+                # For OpenAI/Anthropic async clients
+                if hasattr(underlying, "chat") and hasattr(
+                    underlying.chat, "completions"
+                ):
+                    if hasattr(underlying.chat.completions, "create"):
+                        if inspect.iscoroutinefunction(
+                            underlying.chat.completions.create
+                        ):
+                            return True
+
             # Check if this is an async client by checking for a coroutine method
-            if hasattr(self.client.chat.completions, "create"):
-                return inspect.iscoroutinefunction(self.client.chat.completions.create)
+            if hasattr(self.client, "chat") and hasattr(
+                self.client.chat, "completions"
+            ):
+                if hasattr(self.client.chat.completions, "create"):
+                    return inspect.iscoroutinefunction(
+                        self.client.chat.completions.create
+                    )
+
+            # For instructor-wrapped clients, also check the closure of create_fn
+            # This handles cases where the underlying client is stored in a closure
+            if (
+                hasattr(self.client, "create_fn")
+                and hasattr(self.client.create_fn, "__closure__")
+                and self.client.create_fn.__closure__
+            ):
+                for cell in self.client.create_fn.__closure__:
+                    try:
+                        obj = cell.cell_contents
+                        # Check if the closure object is an async client
+                        if hasattr(obj, "chat") and hasattr(obj.chat, "completions"):
+                            if hasattr(obj.chat.completions, "create"):
+                                if inspect.iscoroutinefunction(
+                                    obj.chat.completions.create
+                                ):
+                                    return True
+                        # Also check for acompletion (e.g., litellm Router)
+                        if hasattr(obj, "acompletion"):
+                            if inspect.iscoroutinefunction(obj.acompletion):
+                                return True
+                    except (ValueError, AttributeError):
+                        # cell_contents might not be accessible
+                        pass
+
             return False
         except (AttributeError, TypeError):
             return False

--- a/src/ragas/llms/litellm_llm.py
+++ b/src/ragas/llms/litellm_llm.py
@@ -45,19 +45,61 @@ class LiteLLMStructuredLLM(InstructorBaseRagasLLM):
         self.is_async = self._check_client_async()
 
     def _check_client_async(self) -> bool:
-        """Determine if the client is async-capable."""
+        """Determine if the client is async-capable.
+
+        Handles multiple cases:
+        1. Direct async clients (e.g., litellm Router with acompletion)
+        2. Instructor-wrapped AsyncInstructor clients
+        3. Instructor-wrapped Instructor clients (need to check underlying client)
+        """
         try:
-            # Check for async completion method
+            # Check if this is an AsyncInstructor wrapper (instructor.AsyncInstructor)
+            if self.client.__class__.__name__ == "AsyncInstructor":
+                return True
+
+            # Check for direct async completion method (e.g., litellm Router)
             if hasattr(self.client, "acompletion"):
-                return inspect.iscoroutinefunction(self.client.acompletion)
-            # Check for async chat completion
+                is_coroutine = inspect.iscoroutinefunction(self.client.acompletion)
+                if is_coroutine:
+                    return True
+
+            # Check for async chat completion (works with instructor-wrapped OpenAI clients)
             if hasattr(self.client, "chat") and hasattr(
                 self.client.chat, "completions"
             ):
                 if hasattr(self.client.chat.completions, "create"):
-                    return inspect.iscoroutinefunction(
-                        self.client.chat.completions.create
-                    )
+                    if inspect.iscoroutinefunction(self.client.chat.completions.create):
+                        return True
+
+            # For instructor-wrapped sync clients that wrap async underlying clients,
+            # check if the wrapped client has async methods
+            if hasattr(self.client, "client"):
+                # This is an instructor-wrapped client, check the underlying client
+                underlying = self.client.client
+                if hasattr(underlying, "acompletion"):
+                    is_coroutine = inspect.iscoroutinefunction(underlying.acompletion)
+                    if is_coroutine:
+                        return True
+
+            # For instructor-wrapped clients, also check the closure of create_fn
+            # This handles cases where the underlying client is stored in a closure
+            # (e.g., when instructor.from_litellm wraps a litellm Router)
+            if (
+                hasattr(self.client, "create_fn")
+                and hasattr(self.client.create_fn, "__closure__")
+                and self.client.create_fn.__closure__
+            ):
+                for cell in self.client.create_fn.__closure__:
+                    try:
+                        obj = cell.cell_contents
+                        # Check if the closure object has acompletion (e.g., litellm Router)
+                        if hasattr(obj, "acompletion"):
+                            if inspect.iscoroutinefunction(obj.acompletion):
+                                return True
+                    except (ValueError, AttributeError):
+                        # cell_contents might not be accessible, or object might not have acompletion
+                        pass
+
             return False
         except (AttributeError, TypeError):
             return False

